File: src/ardrone_follow.cpp
#include<ros/ros.h>
#include<geometry_msgs/Point.h>
#include<geometry_msgs/Twist.h>
#include<ardrone_autonomy/Navdata.h>
#include<ardrone_autonomy/LedAnim.h>
#include<std_msgs/Empty.h>
#include<math.h>

#define SCALE 5

class PidControl
{
  double setPointMax_;
  double setPointMin_;
  double Kp_;
  double Ti_;
  double Td_;

  double outputLimit_;

  double preError_;
  double integral_;

public:
  PidControl( double Kp, double Ti, double Td, double outputLimit ) :
    Kp_( Kp ), Ti_( Ti ), Td_( Td ), outputLimit_( outputLimit ),
    preError_( 0.0 ), integral_( 0.0 )
  {
  }

  void setSetPoint( double setPoint )
  {
    setPointMax_ = setPointMin_ = setPoint;
  }

  void setSetPointMax( double setPointMax )
  {
    setPointMax_ = setPointMax;
  }

  void setSetPointMin( double setPointMin )
  {
    setPointMin_ = setPointMin;
  }

  void setKp( double Kp )
  {
    Kp_ = Kp;
  };

  void setTd( double Td )
  {
    Td_ = Td;
  }

  void setTi( double Ti )
  {
    Ti_ = Ti;
  }

  double getOutput( double PV, ros::Duration Dt )
  {
    double error;

    if( PV < setPointMax_ && PV > setPointMin_ )
      error = 0;
    else
      {
	double errorFromMax = setPointMax_ - PV;
	double errorFromMin = setPointMin_ - PV;

	if( fabs( errorFromMax ) > fabs( errorFromMin ) )
	  error = errorFromMin;
	else
	  error = errorFromMax;
      }

    integral_ += error * Dt.toSec();
    double derivative = ( error - preError_ ) / Dt.toSec();
    preError_ = error;

    double TiRecip = ( Ti_ == 0.0 ) ? 0 : 1/Ti_;
    double output = Kp_ * ( error + TiRecip * integral_ + Td_ * derivative );

    double limitedOutput = fmin( fmax( output, -outputLimit_ ),
				 outputLimit_ );
    return( limitedOutput );
  }
};

class ArdroneFollow
{
  ros::NodeHandle nh_;
  geometry_msgs::Twist current_cmd_;
  ros::Subscriber navdata_;
  ros::Subscriber tracker_;
  ros::Publisher cmd_vel_;

  ros::Timer timer_;
  ros::Timer hoverTimer_;

  ros::Publisher land_pub_;
  ros::Publisher reset_pub_;
  ros::Publisher takeoff_pub_;

  geometry_msgs::Point found_point_;
  geometry_msgs::Point set_point_;

  PidControl xPid, yPid, zPid;

  ros::ServiceClient ledService_;
  int lastAnim_;

public:
  ArdroneFollow()
    : current_cmd_(),
      xPid( 0.020, 0.0, 0.0, 100.0 ), // pos
      yPid( 0.020, 0.0, 0.0, 100.0 ), // pos
      zPid( 0.500, 0.0, 0.0, 50.0 ), // pos
      lastAnim_( -1 )
  {
    navdata_ = nh_.subscribe( "ardrone/navdata", 1000, &ArdroneFollow::navdataCb, this );
    tracker_ = nh_.subscribe( "ardrone_tracker/found_point", 1, &ArdroneFollow::foundpointCb, this );
    cmd_vel_ = nh_.advertise<geometry_msgs::Twist>( "cmd_vel", 1 );
    timer_ = nh_.createTimer( ros::Duration( 0.1 ), &ArdroneFollow::timerCb, this );

    land_pub_ = nh_.advertise<std_msgs::Empty>("ardrone/land", 1);
    takeoff_pub_ = nh_.advertise<std_msgs::Empty>("ardrone/takeoff", 1);
    reset_pub_ = nh_.advertise<std_msgs::Empty>("ardrone/reset", 1);

    xPid.setSetPointMin( 40.0 );
    xPid.setSetPointMax( 60.0 );
    yPid.setSetPointMin( 50.0 );
    yPid.setSetPointMax( 65.0 );
    zPid.setSetPointMin( 40.0 );
    zPid.setSetPointMax( 60.0 );

    ledService_ = nh_.serviceClient<ardrone_autonomy::LedAnim>( "ardrone/setledanimation" );
  }

  ~ArdroneFollow()
  {
  }

  void setLedAnim( int type, int freq = 10 )
  {
    if ( lastAnim_ == type )
      return;

    ardrone_autonomy::LedAnim anim;
    lastAnim_ = anim.request.type = type;
    anim.request.freq = 10;
    anim.request.duration = 3600;

    ledService_.call( anim );
  }

  void takeoff( void )
  {
    takeoff_pub_.publish( std_msgs::Empty() );
    setLedAnim( 9 ); // RED_SNAKE
  }

  void land( void )
  {
    land_pub_.publish( std_msgs::Empty() );
  }

  void reset( void )
  {
    reset_pub_.publish( std_msgs::Empty() );
    setLedAnim( 6 ); // STANDARD
  }

  void navdataCb( const ardrone_autonomy::Navdata &navdata )
  {
    // For now ignore the nav data
  }

  void foundpointCb( const geometry_msgs::Point &found_point )
  {
    found_point_ = found_point;
  }

  void hover( void )
  {
    geometry_msgs::Twist hoverCmd;
    cmd_vel_.publish( hoverCmd );
  }

  void hoverCmdCb( const ros::TimerEvent &event )
  {
    this->hover();
  }

  void timerCb( const ros::TimerEvent &event )
  {
    ros::Duration Dt = event.current_real - event.last_real;
 
    if( found_point_.z == -1.0 )
      {
	// Rotate
	current_cmd_ = geometry_msgs::Twist();
	//	current_cmd_.angular.z = 1.0 / SCALE;
	//current_cmd_.angular.z = 0.0

	setLedAnim( 0, 2 );
      }
    else
      {
	// Run the pid controller
	current_cmd_.angular.z = xPid.getOutput( found_point_.x, Dt );
	current_cmd_.linear.z = yPid.getOutput( found_point_.y, Dt );
	current_cmd_.linear.x = zPid.getOutput( found_point_.z, Dt );

	setLedAnim( 8, 2 );
      }

    cmd_vel_.publish( current_cmd_ );

    // then in 0.5s, publish a hover cmd
    hoverTimer_ = nh_.createTimer( ros::Duration( Dt.toSec() / 2.0 ), &ArdroneFollow::hoverCmdCb, this, /* oneshot = */ true );
  }
};

int main( int argc, char **argv )
{
  ros::init( argc, argv, "ardrone_follow" );
  ArdroneFollow af;

  puts( "reset" );
  af.reset();
  sleep( 2 );
  puts( "takeoff" );
  af.takeoff();

  // clear commands
  af.hover();

  // let it settle after taking off
  sleep( 10 );

  ros::AsyncSpinner spinner(2);

  // Launch
  spinner.start();
  sleep( 300 ); // 5 minutes

  // Land
  puts( "land" );
  af.land();

  ros::shutdown();
  ros::waitForShutdown(); // kind of unnecessary but just in case
  return 0;
}

File: nodes/ardrone_follow.py
#!/usr/bin/env python
import roslib
roslib.load_manifest('ardrone_autonomy')

import sys
import rospy
import math
import pid
import cv2
import numpy as np

from geometry_msgs.msg import Point
from geometry_msgs.msg import Twist
from std_msgs.msg import Empty
from sensor_msgs.msg import Joy
from ardrone_autonomy.msg import Navdata
from ardrone_autonomy.srv import LedAnim

class ArdroneFollow:
    def __init__( self ):
        self.navdata_sub = rospy.Subscriber( "ardrone/navdata", Navdata, 
                                             self.navdata_cb )
        self.tracker_sub = rospy.Subscriber( "ardrone_tracker/found_point",
                                             Point, self.found_point_cb )
        self.cmd_vel_pub = rospy.Publisher( "cmd_vel", Twist )
        self.timer = rospy.Timer( rospy.Duration( 0.25 ), self.timer_cb, False )

        self.land_pub = rospy.Publisher( "ardrone/land", Empty )
        self.takeoff_pub = rospy.Publisher( "ardrone/takeoff", Empty )
        self.reset_pub = rospy.Publisher( "ardrone/reset", Empty )

        self.xPid = pid.Pid( 0.020, 0.0, 0.0, 100.0 )
        self.yPid = pid.Pid( 0.020, 0.0, 0.0, 100.0 )
        self.zPid = pid.Pid( 0.500, 0.0, 0.0,  50.0 )

        self.xPid.setPointMin = 40
        self.xPid.setPointMax = 60

        self.yPid.setPointMin = 50
        self.yPid.setPointMax = 65

        self.zPid.setPointMin = 17
        self.zPid.setPointMax = 23

        self.led_service = rospy.ServiceProxy( "ardrone/setledanimation", LedAnim )
        self.lastAnim = -1

        self.foundpoint = Point( 0, 0, -1 )
        self.old_cmd = self.current_cmd = Twist()

        self.joy_sub = rospy.Subscriber( "joy", Joy, self.callback_joy )
        self.manual_cmd = False
        self.auto_cmd = False

        cv2.namedWindow( 'Follow Clues', cv2.cv.CV_WINDOW_NORMAL )


    def increase_z_setpt( self ):
        self.zPid.setPointMin *= 1.01
        self.zPid.setPointMax *= 1.01

    def decrease_z_setpt( self ):
        self.zPid.setPointMin /= 1.01
        self.zPid.setPointMax /= 1.01

    def callback_joy( self, data ):
        empty_msg = Empty()

        if data.buttons[12] == 1 and self.last_buttons[12] == 0:
            self.takeoff()

        if data.buttons[14] == 1 and self.last_buttons[14] == 0:
            self.land()

        if data.buttons[13] == 1 and self.last_buttons[13] == 0:
            self.reset()

        if data.buttons[15] == 1 and self.last_buttons[15] == 0:
            self.auto_cmd = not self.auto_cmd
            self.hover()

        if data.buttons[4] == 1:
            self.increase_z_setpt()

        if data.buttons[6] == 1:
            self.decrease_z_setpt()

        self.last_buttons = data.buttons

        # Do cmd_vel
        self.current_cmd = Twist()

        self.current_cmd.angular.x = self.current_cmd.angular.y = 0
        self.current_cmd.angular.z = data.axes[2]

        self.current_cmd.linear.z = data.axes[3]
        self.current_cmd.linear.x = data.axes[1]
        self.current_cmd.linear.y = data.axes[0]

        if ( self.current_cmd.linear.x == 0 and
             self.current_cmd.linear.y == 0 and
             self.current_cmd.linear.z == 0 and
             self.current_cmd.angular.z == 0 ):
            self.manual_cmd = False
        else:
            self.setLedAnim( 9 )
            self.manual_cmd = True

        self.cmd_vel_pub.publish( self.current_cmd )

    def setLedAnim( self, animType, freq = 10 ):
        if self.lastAnim == type:
            return

        msg = LedAnim();
        msg.type = animType;
        msg.freq = freq;
        msg.duration = 3600;
        
        self.led_service( type = animType, freq = freq, duration = 255 )
        self.lastAnim = type

    def takeoff( self ):
        self.takeoff_pub.publish( Empty() )
        self.setLedAnim( 9 )

    def land( self ):
        self.land_pub.publish( Empty() )

    def reset( self ):
        self.reset_pub.publish( Empty() )

    def navdata_cb( self, data ):
        pass

    def found_point_cb( self, data ):
        self.found_point = data

    def hover( self ):
        hoverCmd = Twist()
        self.cmd_vel_pub.publish( hoverCmd )

    def hover_cmd_cb( self, data ):
        self.hover()

    def draw_picture( self ):
        vis = np.zeros( ( 360, 640, 3 ), np.uint8);

        cx_min = int(self.xPid.setPointMin * 640 / 100)
        cx_max = int(self.xPid.setPointMax * 640 / 100)
        cx = int( ( cx_min + cx_max ) / 2 )
        
        cy_min = int(self.yPid.setPointMin * 480 / 100)
        cy_max = int(self.yPid.setPointMax * 480 / 100)
        cy = int( ( cy_min + cy_max ) / 2 )

        cv2.rectangle( vis, ( cx_min, cy_min ), ( cx_max, cy_max ),
                       ( 0, 255, 255 ) )

        areasqrt = np.sqrt( 640 * 480 )
        side_min_half = int( self.zPid.setPointMin * areasqrt / 100 / 2 )
        side_max_half = int( self.zPid.setPointMax * areasqrt / 100 / 2 )

        cv2.rectangle( vis, ( cx - side_min_half, cy - side_min_half ),
                       ( cx + side_min_half, cy + side_min_half ),
                       ( 255, 255, 0 ) )

        cv2.rectangle( vis, ( cx - side_max_half, cy - side_max_half ),
                       ( cx + side_max_half, cy + side_max_half ),
                       ( 255, 255, 0 ) )

        cv2.line( vis, ( 320, 180 ), ( int( 320 - self.current_cmd.angular.z * 320 ),
                                       int( 180 - self.current_cmd.linear.z * 180  ) ),
                  ( 0, 255, 0 ),
                  min( max( int( abs( self.current_cmd.linear.x ) * 255 ), 0 ), 255 ) )

        cv2.imshow( 'Follow Clues', vis )
        cv2.waitKey( 1 )

    def timer_cb( self, event ):
        self.draw_picture()

        if event.last_real == None:
            dt = 0
        else:
            dt = ( event.current_real - event.last_real ).to_sec()

        if self.found_point.z == -1.0:
            self.current_cmd = Twist()
            self.setLedAnim( 0, 2 )
        else:
            self.current_cmd.angular.z = self.xPid.get_output( self.found_point.x,
                                                               dt )
            self.current_cmd.linear.z = self.yPid.get_output( self.found_point.y,
                                                              dt )
            self.current_cmd.linear.x = self.zPid.get_output( self.found_point.z,
                                                              dt )

            self.setLedAnim( 8, 2 )

        if self.auto_cmd == False or self.manual_cmd == True:
            self.setLedAnim( 9 )
            return

        self.cmd_vel_pub.publish( self.current_cmd )
        self.hover_timer = rospy.Timer( rospy.Duration( dt / 2.0 ), self.hover_cmd_cb,
                                        True )



def main():
    rospy.init_node( 'ardrone_follow' )
    af = ArdroneFollow()

    try:
        rospy.spin()
    except KeyboardInterrupt:
        print "Keyboard interrupted"

if __name__ == '__main__':
    main()
